plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,138 @@
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_FILESYSTEM_ERROR;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_S3_CROSS_REGION_UNSUPPORTED;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftUnloadSplitSource
+        implements ConnectorSplitSource
+{
+    private static final Logger log = Logger.get(RedshiftUnloadSplitSource.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+
+    private final CompletableFuture<Boolean> resultSetFuture;
+    private final String unloadOutputPath;
+    private final TrinoFileSystem fileSystem;
+
+    private Optional<Location> manifestLocation;
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(ExecutorService executor, Connection connection, PreparedStatement statement, String unloadOutputPath, TrinoFileSystem fileSystem)
+    {
+        requireNonNull(executor, "executor is null");
+        requireNonNull(connection, "connection is null");
+        requireNonNull(statement, "statement is null");
+        this.unloadOutputPath = requireNonNull(unloadOutputPath, "unloadOutputPath is null");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        resultSetFuture = CompletableFuture.supplyAsync(() -> {
+            log.debug("Executing: %s", statement);
+            try {
+                // Exclusively set readOnly to false to avoid query failing with "ERROR: transaction is read-only".
+                connection.setReadOnly(false);
+                return statement.execute();
+            }
+            catch (SQLException e) {
+                if (e instanceof RedshiftException && e.getMessage().contains("The S3 bucket addressed by the query is in a different region from this cluster")) {
+                    throw new TrinoException(REDSHIFT_S3_CROSS_REGION_UNSUPPORTED, "Redshift cluster and S3 bucket in different regions is not supported", e);
+                }
+                throw new RuntimeException(e);
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        return resultSetFuture
+                .thenApply(_ -> {
+                    ConnectorSplitBatch connectorSplitBatch = new ConnectorSplitBatch(getUnloadedFilePaths().stream()
+                            .map(fileInfo -> (ConnectorSplit) new RedshiftUnloadSplit(fileInfo.path, fileInfo.size))
+                            .collect(toImmutableList()), true);
+                    finished = true;
+                    return connectorSplitBatch;
+                });
+    }
+
+    private List<FileInfo> getUnloadedFilePaths()
+    {
+        ImmutableList.Builder<FileInfo> unloadedFilePaths = ImmutableList.builder();
+        manifestLocation = Optional.of(Location.of(unloadOutputPath + "manifest"));
```

Comment:
Check that `unloadOutputPath` ends with `/`.

Reply:
added `verify(unloadOutputPath.endsWith("/"), "unloadOutputPath must end with '/'.");` in constructor of this class.